queries/tags.scm

```diff
@@ -0,0 +1,66 @@
+; Definitions
+
+; * modules
+(call
+  target: (identifier) @ignore
+  (arguments (alias) @name)
+  (#eq? @ignore "defmodule")) @definition.module
+
+; * protocols
+(call
+  target: (identifier) @ignore
+  (arguments (alias) @name)
+  (#eq? @ignore "defprotocol")) @definition.interface
+
+; * implementations for protocols
+(call
+  target: (identifier) @ignore
+  (arguments (alias) @name)
+  (#eq? @ignore "defimpl")) @definition.implementation
+
+; * functions/macros
+(call
+  target: (identifier) @ignore
+  (arguments
+    [
+      ; zero-arity functions with no parentheses
+      (identifier) @name
+      ; regular function clause
+      (call target: (identifier) @name)
+      ; function clause with a guard clause
+      (binary_operator
+        left: (call target: (identifier) @name)
+        operator: "when")
+    ])
+  (#match? @ignore "^(def|defp|defdelegate|defguard|defguardp|defmacro|defmacrop|defn|defnp)$")) @definition.function
+
+; References
+
+; ignore calls to kernel/special-forms keywords
+(call
+  target: (identifier) @ignore
+  (#match? @ignore "^(def|defp|defdelegate|defguard|defguardp|defmacro|defmacrop|defn|defnp|defmodule|defprotocol|defimpl|defstruct|defexception|defoverridable|alias|case|cond|else|for|if|import|quote|raise|receive|require|reraise|super|throw|try|unless|unquote|unquote_splicing|use|with)$"))
+
+; ignore module attributes
+(unary_operator
+  operator: "@"
+  operand: (call
+    target: (identifier) @ignore))
+
+; * function call
+(call
+  target: [
+   ; local
+   (identifier) @name
+   ; remote
+   (dot
+     right: (identifier) @name)
+  ]) @reference.call
+
+; * pipe into function call
+(binary_operator
+  operator: "|>"
+  right: (identifier) @name) @reference.call
+
+; * modules
+(alias) @name @reference.call
```

Comment:
Do we intentionally use `@reference.call` for all references or could we use `@reference.module`?

Reply:
Interestingly the [Python queries](https://github.com/tree-sitter/tree-sitter-python/blob/24b530ca158d2782ea9046e756057a412e16b52f/queries/tags.scm) have neither of those for classes/modules and yet it works :thinking:

Reply:
Good question! Yep I tried out `@reference.module` initially but I thought it was not working. I just gave it another try and  the above output changes to:

```
Mint.MixProject	| module  	def (0, 10) - (0, 25) `defmodule Mint.MixProject do`
Mix.Project	| module  	ref (1, 6) - (1, 17) `use Mix.Project`
project   	| function	def (6, 6) - (6, 13) `def project do`
Mix       	| module  	ref (11, 23) - (11, 26) `start_permanent: Mix.env() == :prod,`
env       	| call    	ref (11, 27) - (11, 30) `start_permanent: Mix.env() == :prod,`
elixirc_paths	| call    	ref (12, 21) - (12, 34) `elixirc_paths: elixirc_paths(Mix.env()),`
Mix       	| module  	ref (12, 35) - (12, 38) `elixirc_paths: elixirc_paths(Mix.env()),`
env       	| call    	ref (12, 39) - (12, 42) `elixirc_paths: elixirc_paths(Mix.env()),`
deps      	| call    	ref (13, 12) - (13, 16) `deps: deps(),`
package   	| call    	ref (29, 15) - (29, 22) `package: package(),`
application	| function	def (46, 6) - (46, 17) `def application do`
Mint.Application	| module  	ref (49, 12) - (49, 28) `mod: {Mint.Application, []}`
package   	| function	def (53, 7) - (53, 14) `defp package do`
elixirc_paths	| function	def (60, 7) - (60, 20) `defp elixirc_paths(:test), do: ["lib", "test/support"]`
elixirc_paths	| function	def (61, 7) - (61, 20) `defp elixirc_paths(_env), do: ["lib"]`
deps      	| function	def (64, 7) - (64, 11) `defp deps do`
```

which looks happy.

I see in some other queries now that there is a little variance like in these c-sharp tags.scm: https://github.com/tree-sitter/tree-sitter-c-sharp/blob/3104df21065af0f3d51e05a96cd0e2ff16a6f982/queries/tags.scm, which also makes me think that what is currently being tagged as `@definition.implementation` should actually be `@reference.interface` for code like

```elixir
defimpl Jason.Encoder, for: List do
```

On the other hand, the ruby query has [this line](https://github.com/tree-sitter/tree-sitter-ruby/blob/888e2e563ed3b43c417f17e57f7e29c39ce9aeea/queries/tags.scm#L61):

```scheme
  [(identifier) (constant)] @name @reference.call
```

which tags modules/classes as `@reference.call`s though and I've seen module references work in ruby code (for example [Appsignal](https://github.com/appsignal/appsignal-ruby/blob/fb9949bbff13606195adb3d6576de32dfd734483/lib/appsignal.rb)), so I'm not sure what the right move is. Maybe all definitions/references are pooled together when performing the search? Or maybe it just helps with ranking?

I think I'll update this line and the `@reference.interface` as well :+1:

Reply:
Ah, `(constant)` is the inner node for both class and module, I missed initially!

If the `Jason.Encoder` alias appears anywhere else we won't know if it's a module or a protocol. I think we should just use `module` for protocols as well (which is actually true), unless I'm missing some benefits of tagging them separately.

Reply:
ah yeah I think that makes sense, I could see things being ambiguous if you do a `@derive {Jason.Encoder, ..}` and search doesn't match up that `@reference.module` with the `@definition.interface`

I'll cut out all that interface stuff in preference of modules :+1:

Reply:
Also, `definition.function` is not consistent with `reference.call` (as opposed to `reference.function`), but we're probably missing some information here, so let's wait ^^